### PR TITLE
Resolve #8696

### LINKED ISSUE #8696: STM retry/orElse semantics
This issue is a case of "behaves differently from Haskell", ~~so I'm not even sure this is a bug.~~ (Although the Haskell behavior seems more useful to me.)

Apparently, Haskell's STM retries a transaction if any of the variables read in any branch of an `orElse` change. (See [section 3.4 here](https://cs.uwaterloo.ca/~Brecht/courses/702/Possible-Readings/transactional-memory/composable-mem-trans-ppopp-2005.pdf).) It seems ZSTM doesn't do this: it only retries if the `TRef` read by the RHS/second transaction changes. For example, this program hangs:

```scala
    for {
      ref1 <- TRef.makeCommit(0)
      ref2 <- TRef.makeCommit(0)
      txn1 = ref1.get.flatMap {
        case 0 => STM.retry
        case n => STM.succeed(n)
      } orElse ref2.get.flatMap {
        case 0 => STM.retry
        case n => STM.succeed(n)
      }
      txn2 = ref1.set(1) // if this is `ref2.set(1)`, it doesn't hang
      fib <- txn1.commit.forkDaemon
      _ <- ZIO.sleep(zio.durationInt(1).second)
      _ <- txn2.commit
      r <- fib.join
    } yield ()
```

### PR DESCRIPTION
**Pull request by BijenderKumar1**
This PR resolves #8696, which happens as the journal reset does not happen accurately, and instead empties the journal. I resolved this issue with the `prepareJournalReset` function.

I tested using the same code provided in the issue #8696, and it no longer hangs.
Thanks!

/claim #8696

### CODE REVIEW

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 1]
```diff
@@ -1621,15 +1621,17 @@ object ZSTM {
      * Creates a function that can reset the journal.
      */
     def prepareResetJournal(journal: Journal): () => Any = {
```
Comment (hearnadam) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala:1623]: The extra set of braces can now be removed.

Comment (BijenderKumar1) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1623]: Sure, Thanks. I will do that.

Comment (BijenderKumar1) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala:1623]: I just added a new commit removing the extra set of braces. Thanks!

**Comment by guizmaii** [comment]
@BijenderKumar1 Could you add some tests, please?

**Comment by BijenderKumar1** [comment]
@guizmaii @hearnadam I have added a new test to make sure that the variable changes on the LHS of an orElse construct, trigger a retry. I have also made sure that all the existing tests pass. Thanks!

File: core/shared/src/main/scala/zio/stm/ZSTM.scala [hunk 2]
```diff
@@ -2051,6 +2066,18 @@ object ZSTM {
         _isChanged = self.isChanged
       }
 
+      /**
+       * Resets the Entry with a given value.
+       */
+      def reset(resetValue: Any): Entry = new Entry {
```
Comment (jdegoes) [inline core/shared/src/main/scala/zio/stm/ZSTM.scala]: This cannot be public.

Comment (BijenderKumar1) [reply core/shared/src/main/scala/zio/stm/ZSTM.scala]: @jdegoes Thanks for the review! I have incorporated it and pushed the changes. Please review it, whenever you are available. Thanks!

**Comment by BijenderKumar1** [comment]
@jdegoes @hearnadam Just a gentle reminder! Thanks!

**Comment by BijenderKumar1** [comment]
@jdegoes @hearnadam Just a gentle reminder to review this. Sorry to bother you. 
Thanks!

**Review by jdegoes — APPROVED** [review]

**Comment by BijenderKumar1** [comment]
@jdegoes Thanks for approving the changes. Just a gentle reminder to please merge the PR as well, if possible? Thanks! 🙏

**Pull request merged**